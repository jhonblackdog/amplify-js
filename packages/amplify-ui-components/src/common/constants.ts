// Namespace
export const AMPLIFY_UI_PREFIX = 'amplify-ui';

// Classnames
export const AMPLIFY_UI_TEXT_INPUT = `${AMPLIFY_UI_PREFIX}-text-input`;
export const AMPLIFY_UI_SCENE = `${AMPLIFY_UI_PREFIX}-scene`;
export const AMPLIFY_UI_SCENE_LOADING = `${AMPLIFY_UI_PREFIX}-scene-loading`;

// Dictionaries

/*
  The below constants will be moved to an i18n solution.
  Storing here to have a common place to pull from once we pick up that work.
*/
// Username common constants
export const USERNAME_SUFFIX = 'username';
export const USERNAME_LABEL = 'Username *';
export const USERNAME_PLACEHOLDER = 'Enter your username';

// Email common constants
export const EMAIL_SUFFIX = 'email';
export const EMAIL_LABEL = 'Email Address *';
export const EMAIL_PLACEHOLDER = 'amplify@example.com';

// Code common constants
export const CODE_SUFFIX = 'code';
export const CODE_LABEL = 'Verification code';
export const CODE_PLACEHOLDER = 'Enter code';
export const SEND_CODE = 'Send Code';
export const CONFIRM = 'Confirm';

// Phone common constants
export const PHONE_SUFFIX = 'phone';
export const PHONE_LABEL = 'Phone Number *';
export const PHONE_PLACEHOLDER = '(555) 555-1212';

// Password common constants
export const FORGOT_PASSWORD_TEXT = 'Forgot your password?';
export const RESET_PASSWORD_TEXT = 'Reset password';
export const RESET_YOUR_PASSWORD = 'Reset your password';
export const PASSWORD_SUFFIX = 'password';
export const PASSWORD_LABEL = 'Password *';
export const PASSWORD_PLACEHOLDER = 'Enter your password';
export const BACK_TO_SIGN_IN = 'Back to Sign In';

// Country Dial Code common constants
export const COUNTRY_DIAL_CODE_SUFFIX = 'country-dial-code-select';
export const COUNTRY_DIAL_CODE_DEFAULT = '+1';

// Sign In common constants
// TODO: Change to SIGN_IN_HEADER_TEXT
export const HEADER_TEXT = 'Sign into your account';
// TODO: Change to SIGN_IN_SUBMIT_BUTTON_TEXT
export const SUBMIT_BUTTON_TEXT = 'Sign in';
export const CREATE_ACCOUNT_TEXT = 'Create account';
export const NO_ACCOUNT_TEXT = 'No account?';
export const CONFIRM_SMS_CODE = 'Confirm SMS Code';
export const CONFIRM_TOTP_CODE = 'Confirm TOTP Code';
export const SIGN_IN_WITH_AMAZON = 'Sign In with Amazon';
export const SIGN_IN_WITH_AUTH0 = 'Sign In with Auth0';
export const SIGN_IN_WITH_FACEBOOK = 'Sign In with Facebook';
export const SIGN_IN_WITH_GOOGLE = 'Sign In with Google';
export const SIGN_IN_WITH_AWS = 'Sign in with AWS';

// Sign Up common constants
export const SIGN_UP_HEADER_TEXT = 'Create a new account';
export const SIGN_UP_SUBMIT_BUTTON_TEXT = 'Create account';
export const HAVE_ACCOUNT_TEXT = 'Have an account?';
export const SIGN_IN_TEXT = 'Sign in';
export const SIGN_UP_USERNAME_PLACEHOLDER = 'Create a username';
export const SIGN_UP_PASSWORD_PLACEHOLDER = 'Create a password';

// Confirm Sign Up common constants
export const CONFIRM_SIGN_UP_HEADER_TEXT = 'Confirm Sign up';
export const CONFIRM_SIGN_UP_SUBMIT_BUTTON_TEXT = 'Confirm';
export const CONFIRM_SIGN_UP_CODE_LABEL = 'Confirmation Code';
export const CONFIRM_SIGN_UP_CODE_PLACEHOLDER = 'Enter your code';
export const CONFIRM_SIGN_UP_LOST_CODE = 'Lost your code?';
export const CONFIRM_SIGN_UP_RESEND_CODE = 'Resend Code';

// Sign Out
export const SIGN_OUT = 'Sign Out';

// Auth Keys
export const AUTH_SOURCE_KEY = 'amplify-auth-source';

// Error message Common Constants
export const PHONE_EMPTY_ERROR_MESSAGE = 'Phone number can not be empty';
export const NO_AUTH_MODULE_FOUND = 'No Auth module found, please ensure @aws-amplify/auth is imported';
