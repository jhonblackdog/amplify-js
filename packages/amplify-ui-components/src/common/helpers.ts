import { cx } from 'emotion';

// TODO: rename
export const styleNuker = (override: boolean, cn: string, ecn: string): string => override ? cn : cx(ecn, cn);
