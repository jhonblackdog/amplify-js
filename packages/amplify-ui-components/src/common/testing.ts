export const pixelThreshold = 10;
