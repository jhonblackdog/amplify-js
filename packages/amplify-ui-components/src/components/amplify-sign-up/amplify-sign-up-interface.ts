export interface AmplifySignUpFormFooterProps {
  submitButtonText: string;
  haveAcccountText: string;
  signInText: string;
  overrideStyle?: boolean;
  handleAuthStateChange?: any;
}

export interface AmplifySignUpAttributes {
  username: string;
  password: string;
  attributes?: {
    email?: string;
    phone_number?: string;
  };
}

export interface PhoneNumberInterface {
  countryDialCodeValue?: string;
  phoneNumberValue?: string;
}
