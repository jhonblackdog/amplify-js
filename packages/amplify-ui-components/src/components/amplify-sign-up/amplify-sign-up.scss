.sign-up-form-footer {
  font-family: var(--amplify-font-family);
  font-weight: 300;
  font-size: var(--amplify-secondary-font-size);
  color: var(--amplify-grey);
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: baseline;
}
