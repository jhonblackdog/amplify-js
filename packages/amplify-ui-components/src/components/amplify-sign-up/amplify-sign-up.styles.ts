import { css } from 'emotion';

export const signUpFormFooter = css`
  font-family: var(--font-family);
  font-weight: 300;
  font-size: var(--secondary-font-size);
  color: var(--grey);
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: baseline;
`;
