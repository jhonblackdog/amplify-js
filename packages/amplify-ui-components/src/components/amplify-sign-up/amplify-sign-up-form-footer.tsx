import { FunctionalComponent as FC, h } from '@stencil/core';
import { AmplifySignUpFormFooterProps } from './amplify-sign-up-interface';
import { signUpFormFooter } from './amplify-sign-up.styles';

export const AmplifySignUpFormFooter: FC<AmplifySignUpFormFooterProps> = ({
  submitButtonText,
  haveAcccountText,
  signInText,
  overrideStyle = false,
  handleAuthStateChange,
}) => (
  <div class={signUpFormFooter}>
    <span>
      {haveAcccountText} <amplify-link onClick={() => handleAuthStateChange('signin')}>{signInText}</amplify-link>
    </span>
    <amplify-button type="submit" overrideStyle={overrideStyle}>
      {submitButtonText}
    </amplify-button>
  </div>
);
