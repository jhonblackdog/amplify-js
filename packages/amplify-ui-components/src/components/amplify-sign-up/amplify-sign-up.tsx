import { I18n } from '@aws-amplify/core';
import { Auth } from '@aws-amplify/auth';
import { Component, Prop, h, State } from '@stencil/core';
import {
  FormFieldTypes,
  PhoneNumberInterface,
} from '../../components/amplify-auth-fields/amplify-auth-fields-interface';
import {
  PHONE_SUFFIX,
  COUNTRY_DIAL_CODE_DEFAULT,
  COUNTRY_DIAL_CODE_SUFFIX,
  NO_AUTH_MODULE_FOUND,
} from '../../common/constants';
import { AuthState, AuthStateHandler, UsernameAliasStrings } from '../../common/types/auth-types';
import { AmplifySignUpAttributes } from './amplify-sign-up-interface';
import {
  dispatchAuthStateChangeEvent,
  dispatchToastHubEvent,
  composePhoneNumberInput,
  checkUsernameAlias,
} from '../../common/helpers';
import { Translations } from '../../common/Translations';

@Component({
  tag: 'amplify-sign-up',
  styleUrl: 'amplify-sign-up.scss',
  shadow: true,
})
export class AmplifySignUp {
  /** Fires when sign up form is submitted */
  @Prop() handleSubmit: (event: Event) => void = event => this.signUp(event);
  /** Engages when invalid actions occur, such as missing field, etc. */
  @Prop() validationErrors: string;
  /** Used for header text in sign up component */
  @Prop() headerText: string = I18n.get(Translations.SIGN_UP_HEADER_TEXT);
  /** Used for the submit button text in sign up component */
  @Prop() submitButtonText: string = I18n.get(Translations.SIGN_UP_SUBMIT_BUTTON_TEXT);
  /** Used for the submit button text in sign up component */
  @Prop() haveAccountText: string = I18n.get(Translations.SIGN_UP_HAVE_ACCOUNT_TEXT);
  /** Used for the submit button text in sign up component */
  @Prop() signInText: string = I18n.get(Translations.SIGN_IN_TEXT);
  /**
   * Form fields allows you to utilize our pre-built components such as username field, code field, password field, email field, etc.
   * by passing an array of strings that you would like the order of the form to be in. If you need more customization, such as changing
   * text for a label or adjust a placeholder, you can follow the structure below in order to do just that.
   * ```
   * [
   *  {
   *    type: string,
   *    label: string,
   *    placeholder: string,
   *    hint: string | Functional Component | null,
   *    required: boolean
   *  }
   * ]
   * ```
   */
  @Prop() formFields: FormFieldTypes | string[] = [];
  /** Auth state change handler for this component
   * e.g. SignIn -> 'Create Account' link -> SignUp
   */
  @Prop() handleAuthStateChange: AuthStateHandler = dispatchAuthStateChangeEvent;
  /** Username Alias is used to setup authentication with `username`, `email` or `phone_number`  */
  @Prop() usernameAlias: UsernameAliasStrings = 'username';
  // private userInput: string | PhoneNumberInterface;
  private newFormFields: FormFieldTypes | string[] = [];
  private phoneNumber: PhoneNumberInterface = {
    countryDialCodeValue: COUNTRY_DIAL_CODE_DEFAULT,
    phoneNumberValue: null,
  };

  @State() loading: boolean = false;
  @State() signUpAttributes: AmplifySignUpAttributes = {
    username: '',
    password: '',
    attributes: {},
  };

  handleFormFieldInputChange(fieldType) {
    switch (fieldType) {
      case 'username':
        return event => (this.signUpAttributes.username = event.target.value);
      case 'password':
        return event => (this.signUpAttributes.password = event.target.value);
      case 'email':
        return event => (this.signUpAttributes.attributes.email = event.target.value);
      case 'phone_number':
        return event => this.handlePhoneNumberChange(event);
      default:
        return event => (this.signUpAttributes.attributes[fieldType] = event.target.value);
    }
  }

  handleFormFieldInputWithCallback(event, field) {
    const fnToCall = field['handleInputChange'];
    const callback =
      field.type === 'phone_number'
        ? event => (this.signUpAttributes.attributes.phone_number = event.target.value)
        : this.handleFormFieldInputChange(field.type);
    fnToCall(event, callback.bind(this));
  }

  handlePhoneNumberChange(event) {
    const name = event.target.name;
    const value = event.target.value;

    /** Cognito expects to have a string be passed when signing up. Since the Select input is separate
     * input from the phone number input, we need to first capture both components values and combined
     * them together.
     */

    if (name === COUNTRY_DIAL_CODE_SUFFIX) {
      this.phoneNumber.countryDialCodeValue = value;
    }

    if (name === PHONE_SUFFIX) {
      this.phoneNumber.phoneNumberValue = value;
    }
  }

  // TODO: Add validation
  // TODO: Prefix
  async signUp(event: Event) {
    if (event) {
      event.preventDefault();
    }
    if (!Auth || typeof Auth.signUp !== 'function') {
      throw new Error(NO_AUTH_MODULE_FOUND);
    }
    if (this.phoneNumber.phoneNumberValue) {
      this.signUpAttributes.attributes.phone_number = composePhoneNumberInput(this.phoneNumber);
    }
    switch (this.usernameAlias) {
      case 'email':
      case 'phone_number':
        this.signUpAttributes.username = this.signUpAttributes.attributes[this.usernameAlias];
        break;
      case 'username':
      default:
        break;
    }

    try {
      const data = await Auth.signUp(this.signUpAttributes);
      this.handleAuthStateChange(AuthState.ConfirmSignUp, { ...data.user, signUpAttrs: this.signUpAttributes });
    } catch (error) {
      dispatchToastHubEvent(error);
    }
  }

  async componentWillLoad() {
    checkUsernameAlias(this.usernameAlias);
    if (this.formFields.length === 0) {
      switch (this.usernameAlias) {
        case 'email':
          this.newFormFields = [
            {
              type: 'email',
              placeholder: I18n.get(Translations.SIGN_UP_EMAIL_PLACEHOLDER),
              required: true,
              handleInputChange: this.handleFormFieldInputChange('email'),
              inputProps: {
                'data-test': 'sign-up-email-input',
              },
            },
            {
              type: 'password',
              placeholder: I18n.get(Translations.SIGN_UP_PASSWORD_PLACEHOLDER),
              required: true,
              handleInputChange: this.handleFormFieldInputChange('password'),
              inputProps: {
                'data-test': 'sign-up-password-input',
              },
            },
            {
              type: 'phone_number',
              required: true,
              handleInputChange: this.handleFormFieldInputChange('phone_number'),
              inputProps: {
                'data-test': 'sign-up-phone-number-input',
              },
            },
          ];
          break;
        case 'phone_number':
          this.newFormFields = [
            {
              type: 'phone_number',
              required: true,
              handleInputChange: this.handleFormFieldInputChange('phone_number'),
              inputProps: {
                'data-test': 'sign-up-phone-number-input',
              },
            },
            {
              type: 'password',
              placeholder: I18n.get(Translations.SIGN_UP_PASSWORD_PLACEHOLDER),
              required: true,
              handleInputChange: this.handleFormFieldInputChange('password'),
              inputProps: {
                'data-test': 'sign-up-password-input',
              },
            },
            {
              type: 'email',
              placeholder: I18n.get(Translations.SIGN_UP_EMAIL_PLACEHOLDER),
              required: true,
              handleInputChange: this.handleFormFieldInputChange('email'),
              inputProps: {
                'data-test': 'sign-up-email-input',
              },
            },
          ];
          break;
        case 'username':
        default:
          this.newFormFields = [
            {
              type: 'username',
              placeholder: I18n.get(Translations.SIGN_UP_USERNAME_PLACEHOLDER),
              required: true,
              handleInputChange: this.handleFormFieldInputChange('username'),
              inputProps: {
                'data-test': 'sign-up-username-input',
              },
            },
            {
              type: 'password',
              placeholder: I18n.get(Translations.SIGN_UP_PASSWORD_PLACEHOLDER),
              required: true,
              handleInputChange: this.handleFormFieldInputChange('password'),
              inputProps: {
                'data-test': 'sign-up-password-input',
              },
            },
            {
              type: 'email',
              placeholder: I18n.get(Translations.SIGN_UP_EMAIL_PLACEHOLDER),
              required: true,
              handleInputChange: this.handleFormFieldInputChange('email'),
              inputProps: {
                'data-test': 'sign-up-email-input',
              },
            },
            {
              type: 'phone_number',
              required: true,
              handleInputChange: this.handleFormFieldInputChange('phone_number'),
              inputProps: {
                'data-test': 'sign-up-phone-number-input',
              },
            },
          ];
          break;
      }
    } else {
      this.formFields.forEach(field => {
        const newField = { ...field };
        newField['handleInputChange'] = event => this.handleFormFieldInputWithCallback(event, field);
        this.newFormFields.push(newField);
      });
    }
  }

  render() {
    return (
      <amplify-form-section headerText={this.headerText} handleSubmit={this.handleSubmit} testDataPrefix={'sign-up'}>
        <amplify-auth-fields formFields={this.newFormFields} />
        <div class="sign-up-form-footer" slot="amplify-form-section-footer">
          <span>
            {this.haveAccountText}{' '}
            <amplify-button
              variant="anchor"
              onClick={() => this.handleAuthStateChange(AuthState.SignIn)}
              data-test="sign-up-sign-in-link"
            >
              {this.signInText}
            </amplify-button>
          </span>
          <amplify-button type="submit" data-test="sign-up-create-account-button">
            <amplify-loading-spinner style={{ display: this.loading ? 'initial' : 'none' }} />
            <span style={{ display: this.loading ? 'none' : 'initial' }}>{this.submitButtonText}</span>
          </amplify-button>
        </div>
      </amplify-form-section>
    );
  }
}
