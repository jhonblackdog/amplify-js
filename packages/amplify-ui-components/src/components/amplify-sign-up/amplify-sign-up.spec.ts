import { newSpecPage } from '@stencil/core/testing';
import { AmplifySignUp } from './amplify-sign-up';
import { SIGN_UP_HEADER_TEXT, SIGN_UP_SUBMIT_BUTTON_TEXT } from '../../common/constants';

describe('amplify-sign-up spec:', () => {
  describe('Component logic ->', () => {
    let signUp;

    beforeEach(() => {
      signUp = new AmplifySignUp();
    });

    it('should render `handleSubmit` as undefined by default', () => {
      expect(signUp.handleSubmit).toBeDefined();
    });

    it('should render `validationErrors` as undefined by default', () => {
      expect(signUp.validationErrors).toBeUndefined();
    });

    it('should render `headerText` to `Create a new account` by default', () => {
      expect(signUp.headerText).toEqual(SIGN_UP_HEADER_TEXT);
    });

    it('should render `submitButtonText` to `Create account` by default', () => {
      expect(signUp.submitButtonText).toEqual(SIGN_UP_SUBMIT_BUTTON_TEXT);
    });

    it('should render `overrideStyle` to false by default', () => {
      expect(signUp.overrideStyle).toBe(false);
    });
  });
  describe('Render logic ->', () => {
    it('should render a `sign up` form by default', async () => {
      const page = await newSpecPage({
        components: [AmplifySignUp],
        html: `<amplify-sign-up></amplify-sign-up>`,
      });

      expect(page.root).toMatchSnapshot();
    });
  });
});
