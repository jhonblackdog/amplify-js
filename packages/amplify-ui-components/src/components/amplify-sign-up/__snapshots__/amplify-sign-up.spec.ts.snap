// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-sign-up spec: Render logic -> should render a \`sign up\` form by default 1`] = `
<amplify-sign-up>
  <amplify-form-section headertext="Create a new account">
    <amplify-auth-fields></amplify-auth-fields>
    <div slot="amplify-form-section-footer">
      <div class="css-1r8ajg4">
        <span>
          Have an account?
          <amplify-link>
            Sign in
          </amplify-link>
        </span>
        <amplify-button type="submit">
          Create account
        </amplify-button>
      </div>
    </div>
  </amplify-form-section>
</amplify-sign-up>
`;
